app_dart/lib/src/request_handlers/scheduler/backfill_grid.dart

```diff
@@ -0,0 +1,142 @@
+
+import 'package:meta/meta.dart';
+
+import '../../model/ci_yaml/target.dart';
+import '../../service/luci_build_service/opaque_commit.dart';
+import '../../service/luci_build_service/pending_task.dart';
+import '../../service/scheduler/policy.dart';
+
+/// An indexed grid-like mutable view of the last recent N commits and M tasks.
+///
+/// It is the primary input to [BackfillStrategy].
+///
+/// Some common tasks are performed for every grid, independent of a strategy:
+/// - Tasks (& targets) are removed that do not exist at tip-of-tree[^1];
+/// - Tasks (& targets) are removed that do not use backfilling ([BatchPolicy]);
+///
+/// [^1]: The latest commit in the default branch of the current repository.
+final class BackfillGrid {
+  /// Creates an indexed data structure from the provided `(commit, [tasks])`.
+  ///
+  /// Some automatic filtering is applied, removing targets/tasks where:
+  /// - [tipOfTreeTargets] that does not use [BatchPolicy];
+  /// - [tipOfTreeTargets] without a task;
+  /// - task that does have a matching [tipOfTreeTargets].
+  factory BackfillGrid.from(
+    Iterable<(OpaqueCommit, List<OpaqueTask>)> grid, {
+    required Iterable<Target> tipOfTreeTargets,
+  }) {
+    final targetsByName = {for (final t in tipOfTreeTargets) t.name: t};
+    final commitsByName = <String, OpaqueCommit>{};
+    final tasksByName = <String, List<OpaqueTask>>{};
+    for (final (commit, tasks) in grid) {
+      commitsByName[commit.sha] = commit;
+      for (final task in tasks) {
+        // Must exist at ToT (in this Map) and must be BatchPolicy.
+        if (targetsByName[task.name]?.schedulerPolicy is! BatchPolicy) {
+          // Even if it existed, let's remove it at this point because it is no
+          // longer relevant to the BackfillGrid, and if there are future API
+          // changes to the class it shouldn't show up.
+          targetsByName.remove(task.name);
+          continue;
+        }
+        (tasksByName[task.name] ??= []).add(task);
+      }
+    }
+
+    // Final filtering step: remove empty targets/tasks.
+    tasksByName.removeWhere((_, tasks) => tasks.isEmpty);
+    targetsByName.removeWhere((name, _) => !tasksByName.containsKey(name));
+
+    return BackfillGrid._(commitsByName, targetsByName, tasksByName);
+  }
+
+  BackfillGrid._(
+    this._commitsBySha, //
+    this._targetsByName,
+    this._tasksByName,
+  );
+
+  final Map<String, OpaqueCommit> _commitsBySha;
+  final Map<String, List<OpaqueTask>> _tasksByName;
+  final Map<String, Target> _targetsByName;
+
+  /// Returns a [BackfillTask] with the provided LUCI scheduling [priority].
+  @useResult
+  BackfillTask createBackfillTask(OpaqueTask task, {required int priority}) {
+    return BackfillTask._from(
+      task,
+      target: _targetsByName[task.name]!,
+      commit: _commitsBySha[task.commitSha]!,
+      priority: priority,
+    );
+  }
+
+  /// Removes tasks (a column) from the grid that return `true` for [predicate].
```

Comment:
/// Removes a task column from the grid for which [predicate] returns `true`.

Reply:
Done.